### ISSUE TITLE
Secret Server 10.4 support

### ISSUE THREAD

**Issue by jackson5sec** [denandz/SecretServerSecretStealer#2]
Hey, ran into something interesting on a thycotic 10.4 server. (Verified no HSM or DPAPI enabled)
Exception calling "DecryptConfig" with "1" argument(s): "The input stream is not a valid binary format.

Any ideas as to what that means? I tried changing the encoding of the file from ANSI to UTF8 and its still errorring out during the Get-MasterKeys function

**Comment by jackson5sec** [comment]
should probably note this was for the encryption.config file parsing

**Comment by denandz** [comment]
That's unfortunate. I'll have to take a look and see how the new version is handling the master key crypto. Might pay to take a look at the Thycotic.ihawu.Base.FileHydrator class on your instance, and confirm the the key and IV are the same? Maybe they've been rolled. I couldn't find anything in the Secret Server changelog that suggests that's the case, but who knows.

**Comment by jackson5sec** [comment]
That was the fastest response ive ever seen. lol
Is that a file located in the webroot?

**Comment by denandz** [comment]
It'll be in a .dll somewhere. If memory serves correctly you can find them in the inetpub bin folder? You'll need to pull apart the DLL with something like dnspy.

**Comment by jackson5sec** [comment]
Thycotic.ihawu.Base.dll yup it was in webroot\bin. I'll take a look tnight and see if I can find it :) thx for the help!

**Comment by denandz** [comment]
Good luck! If you figure it out can you please let me know? I can likely build some version-checking into the script and have it support the newer versions.

**Comment by jackson5sec** [comment]
Yup totally, me and a coworker are going to look at it now haha

**Comment by jackson5sec** [comment]
// Token: 0x06000010 RID: 16 RVA: 0x00002973 File Offset: 0x00001973
		private string GetKey(string x)
		{
			return "020216980119760c0b79017097830b1d" + x;
		}

		// Token: 0x06000011 RID: 17 RVA: 0x00002980 File Offset: 0x00001980
		private string GetIV(int i)
		{
			return "7a790a22020b6eb3630cdd080310d40a" + i.ToString().Substring(0, i - 1);
		}

**Comment by jackson5sec** [comment]
Was that the same as the one you derived?

**Comment by denandz** [comment]
Yea those are the same, the '+ x' is kinda odd though. Maybe pull out that class and see what happens when you call Load("path_to_encrypted_conf")? Should be able to break on GetKey and GetIV and see whats going on

EDIT: I think either the crypto step is failing for some reason, or they're doing something different with the binary formatter that's causing the error.

**Comment by jackson5sec** [comment]
Scratching my head on this one, do you have a copy of the dll file?

I think it might be the binary formatter, but I don't have another version to compare it to

EDIT: Seeing a weird string, not sure what they are using it for... they also pass an int of 1 to GetIV and then an empty string to GetKey... really weird

**Comment by jackson5sec** [comment]
// Token: 0x0600000A RID: 10 RVA: 0x000025B0 File Offset: 0x000015B0
		public object Load(string fileName)
		{
			this.Garbage2(1L);
			BinaryFormatter formatter = new BinaryFormatter();
			this.Garbage3(1u);
			if (File.Exists(fileName))
			{
				using (FileStream fileStream = File.OpenRead(fileName))
				{
					if (this.Garbage4(1) == this.Garbage("eb36341101223b60200c2d080310d40e", 256))
					{
						return this.Garbage(fileStream, new BinaryFormatter());
					}
					return this.LoadFromStream(formatter, fileStream, 1);
				}
			}
			return null;
		}

		// Token: 0x0600000B RID: 11 RVA: 0x00002634 File Offset: 0x00001634
		public void Save(string fileName, object o)
		{
			this.Garbage1(1);
			this.Garbage2(1L);
			BinaryFormatter formatter = new BinaryFormatter();
			this.Garbage3(1u);
			this.Garbage4(1);
			using (FileStream fileStream = File.Create(fileName))
			{
				if (this.Garbage("eb36341101223b60200c2d080310d40e", 256) == this.Garbage4(1))
				{
					this.Garbage(fileStream, new BinaryFormatter());
				}
				this.SaveToStream(formatter, o, fileStream);
			}
		}

		// Token: 0x0600000C RID: 12 RVA: 0x000026C0 File Offset: 0x000016C0
		private object LoadFromStream(BinaryFormatter formatter, FileStream stream, int garbage)
		{
			this.Garbage5(DateTime.Now);
			AesCryptoServiceProvider aesCryptoServiceProvider = new AesCryptoServiceProvider();
			this.Garbage(aesCryptoServiceProvider, 32);
			this.Garbage6(this.GetIV(1), this.GetIV(1));
			this.Garbage(aesCryptoServiceProvider, 128);
			this.SetKeyAndIV(aesCryptoServiceProvider, "eb36341101223b60200c2d080310d40e");
			this.Garbage(aesCryptoServiceProvider, 256);
			this.Garbage7(1, 2);
			this.Garbage(aesCryptoServiceProvider, 512);
			ICryptoTransform transform = aesCryptoServiceProvider.CreateDecryptor();
			CryptoStream cryptoStream = new CryptoStream(stream, transform, CryptoStreamMode.Read);
			this.Garbage8(true);
			this.Garbage9('c');
			object result = formatter.Deserialize(cryptoStream);
			cryptoStream.Close();
			stream.Close();
			this.Garbage10(1);
			return result;
		}

		// Token: 0x0600000D RID: 13 RVA: 0x00002774 File Offset: 0x00001774
		private void SetKeyAndIV(AesCryptoServiceProvider aes, string s)
		{
			if (this.Garbage("", 1).ToString() == this.Garbage9('\n') || s == "")
			{
				return;
			}
			this.Garbage10(1);
			aes.IV = this.FromHexaDecimal(this.GetIV(1));
			this.Garbage10(1);
			aes.Key = this.FromHexaDecimal(this.GetKey(""));
		}

**Comment by denandz** [comment]
Ok so they've done something new with the encryption.config, check this out:

```
PS C:\inetpub\wwwroot\SecretServer> add-type -path .\bin\Thycotic.ihawu.Base.dll
PS C:\inetpub\wwwroot\SecretServer> add-type -path .\bin\Thycotic.ihawu.Business.dll
PS C:\inetpub\wwwroot\SecretServer> $fh = [Thycotic.ihawu.Base.FileHydrator]::new()
PS C:\inetpub\wwwroot\SecretServer> $fh.Load("C:\inetpub\wwwroot\SecretServer\database.config")

Pairs
-----
{databaseType, sqlServerDatabase, sqlServerLocation, sqlServerUsername...}


PS C:\inetpub\wwwroot\SecretServer> # yay
PS C:\inetpub\wwwroot\SecretServer> $fh.Load("C:\inetpub\wwwroot\SecretServer\encryption.config")
Exception calling "Load" with "1" argument(s): "The input stream is not a valid binary format. The starting contents (in bytes) are:
6A-69-4E-70-10-75-4C-5A-F1-DC-FC-F9-19-26-07-EF-23 ..."
At line:1 char:1
+ $fh.Load("C:\inetpub\wwwroot\SecretServer\encryption.config")
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : SerializationException
```

So... yea. They're still using the FileHydrator for the database config, but they've change the encryption.config loader. It doesn't look like DPAPI is enabled by default, so there is probably another layer of encryption garbage that you'll have to figure out in order to get the master Key and IV out of the encryption.config. 

Why would they do something silly like, I dunno, explain the failure scenario of this software to their users when you can just ADD MORE CRYPTO >_> 

Anyway, the application can get the keys, so you can get the keys, it'll just be a matter of figuring out what they're doing with the encryption.config now. I'm afraid I don't have much time to work on this further right now, but I'll try get around to it in the near future. Let me know if you have any luck getting the encryption.config logic figured out and I'll add it to the powershell script

**Comment by jackson5sec** [comment]
All good, this has turned into more of a learning exercise for me now so no time constraints. Thx for your help!

**Comment by denandz** [comment]
Fixed by #3 10.4 is now supported.
